Model/Indexer/Data.php

```diff
@@ -118,15 +183,75 @@ public function executeRow($id)
      */
     public function execute($ids)
     {
-        $storesWithNosto = $this->nostoHelperAccount->getStoresWithNosto();
-        foreach ($storesWithNosto as $store) {
-            try {
-                $this->nostoServiceIndex->indexProducts($store, $ids);
-                // Catch only MemoryOutOfBoundsException as this is the most expected ones
-                // And the ones we are interested of
-            } catch (MemoryOutOfBoundsException $e) {
-                $this->nostoLogger->error($e->getMessage());
+        $dimensionProvider = $this->getDimensionsProvider();
+        if ($dimensionProvider === null) {
+            $this->executeInSequence($ids);
+        } else {
+            $this->executeInParallel($dimensionProvider, $ids);
+        }
+    }
+
+    /**
+     * @param array $dimensions
+     * @param \Traversable|null $entityIds
+     * @throws NostoException
+     */
+    public function executeByDimensions(array $dimensions, \Traversable $entityIds = null)
+    {
+        if (count($dimensions) > 1 || !isset($dimensions[StoreDimensionProvider::DIMENSION_NAME])) {
+            throw new \InvalidArgumentException('Indexer "' . self::INDEXER_ID . '" support only Store dimension');
+        }
+
+        $storeId = $dimensions[StoreDimensionProvider::DIMENSION_NAME]->getValue();
```

Comment:
Create four functions here and call the process manager to execute it. The logic for creating N functions and queueing them should be here.

Reply:
```
        foreach ($dimensionProvider->getIterator() as $dimension) {
            $userFunctions[] = function () use ($dimension, $ids) {
                /** @var Dimension[] $dimension  */
                $this->executeByDimensions($dimension, new \ArrayIterator($ids));
            };
        }
        if (count($userFunctions) > 1) {
          $this->getProcessManager()->execute($userFunctions);
       } else {
         $userFunctions[0]();
       }
```